=== modified file 'Bugzilla/User.pm'
--- Bugzilla/User.pm	2011-12-28 22:15:49 +0000
+++ Bugzilla/User.pm	2012-01-23 16:15:45 +0000
@@ -125,7 +125,7 @@
         extern_id
         is_enabled
     );
-    push(@cols, 'cryptpassword') if exists $self->{cryptpassword};
+    push(@cols, 'cryptpassword', 'password_changed') if exists $self->{cryptpassword};
     return @cols;
 };
 
@@ -186,6 +186,8 @@
         if (exists $changes->{login_name} || exists $changes->{disabledtext}
             || exists $changes->{cryptpassword});
 
+    $self->set_password_history() if exists $changes->{cryptpassword};
+
     # XXX Can update profiles_activity here as soon as it understands
     #     field names like login_name.
     
@@ -278,7 +280,12 @@
     delete $self->{identity};
 }
 
-sub set_password { $_[0]->set('cryptpassword', $_[1]); }
+sub set_password { 
+    my ($self, $value) = @_;
+    $self->set('cryptpassword', $value);
+    $self->check_password_history($value);
+    $self->{password_changed} = Bugzilla->dbh->selectrow_array('SELECT LOCALTIMESTAMP(0)');
+}
 
 sub set_disabledtext {
     $_[0]->set('disabledtext', $_[1]);
@@ -1852,6 +1859,7 @@
 
     $dbh->bz_start_transaction();
 
+    $_[0]->{password_changed} = $dbh->selectrow_array('SELECT LOCALTIMESTAMP(0)');
     my $user = $class->SUPER::create(@_);
 
     # Turn on all email for the new user

=== modified file 'collectstats.pl'
--- collectstats.pl	2011-11-26 13:18:04 +0000
+++ collectstats.pl	2012-01-24 11:01:21 +0000
@@ -49,6 +49,9 @@
 # Print the help message if that switch was selected.
 pod2usage({-verbose => 1, -exitval => 1}) if $switch{'help'};
 
+BEGIN { Bugzilla->extensions };
+use Bugzilla::Extension::PasswordReset::Util;
+
 # Turn off output buffering (probably needed when displaying output feedback
 # in the regenerate mode).
 $| = 1;
@@ -160,7 +163,9 @@
 # Uncomment the following line for performance testing.
 #print "Total time taken " . delta_time($tstart, $tend) . "\n";
 
-CollectSeriesData();
+#CollectSeriesData();
+
+send_password_notification_email();
 
 sub collect_stats {
     my $dir = shift;

=== modified file 'userprefs.cgi'
--- userprefs.cgi	2012-01-18 21:32:46 +0000
+++ userprefs.cgi	2012-01-23 16:15:45 +0000
@@ -101,11 +101,10 @@
             validate_password($pwd1, $pwd2);
 
             if ($oldpassword ne $pwd1) {
-                my $cryptedpassword = bz_crypt($pwd1);
-                $dbh->do(q{UPDATE profiles
-                              SET cryptpassword = ?
-                            WHERE userid = ?},
-                         undef, ($cryptedpassword, $user->id));
+                $user->set_password($pwd1);
+                $dbh->bz_start_transaction();
+                $user->update();
+                $dbh->bz_commit_transaction();
 
                 # Invalidate all logins except for the current one
                 Bugzilla->logout(LOGOUT_KEEP_CURRENT);

